viewing_party/party.py

```diff
@@ -0,0 +1,204 @@
+    add the movie to the "watched" list inside of user_data
+    '''
+    # verify 'watched' is a key in user_data
+    user_data["watched"].append(movie)
+    return user_data
+    
+def add_to_watchlist(user_data, movie):
+    '''
+    the value of user_data will be a dictionary with a 
+    key "watchlist", and a value which is a list of 
+    dictionaries representing the movies the user wants to watch
+    & add the movie to the "watchlist" list inside of user_data
+    '''
+    user_data["watchlist"].append(movie)
+    return user_data
+
+def watch_movie(user_data, title):
+    '''
+    user_data: dictionary with a "watchlist" and a "watched"
+    title: string that represents the title of the movie 
+        the user has watched
+    if movie in watchlist, remove from watchlist, add to watched
+    returns user_data
+    '''
+    for movie in user_data["watchlist"]: 
+        if title == movie["title"]:
+            user_data["watchlist"].remove(movie)
+            user_data["watched"].append(movie)
+    return user_data
+
+def get_watched_avg_rating(user_data):
+    num_movies = len(user_data["watched"])
+    if num_movies > 0:
+        sum = 0
+        for movie in user_data["watched"]:
+            sum += movie["rating"]
+        avg = sum / num_movies
+    else:
+        avg = 0.0
+    return avg
+
+
+def get_most_watched_genre(user_data): 
+    '''
+    determines which genre is most frequently occurring
+    return most frequently watched
+    if not applicable, return None
+    '''
+    # if there are watched movies, count genre freq
+    if len(user_data["watched"])> 0:
+        # store counters in genres
+        genres = dict()
+        for movie in user_data["watched"]:
+            if movie["genre"] in genres:
+                genres[movie["genre"]] += 1
+            else:
+                genres[movie["genre"]] = 1
+        # get max genre freq
+        return max(genres, key=genres.get) 
+    else:
+        return None
+
+def movie_list(person):
+    '''
+    return list of a users watched movies
+    '''
+    persons_movies = [movie["title"] for movie in person["watched"]]
+    return persons_movies
+
+def friends_watched_movies(user_data):
+    '''
+    return set of all friends movies combined
+    '''
+    friends_movie_list = []
+    for friend in user_data["friends"]:
+        friends_movie_list+=movie_list(friend)
+    friends_movie_set = set(friends_movie_list)
+    return friends_movie_set
+
+def get_unique_watched(user_data):
+    ''' 
+    return movies user has watched but no friend has
+    '''
+    # get users watched movies
+    user_movies = set(movie_list(user_data))
+    # get friends watched movies
+    friends_movie_set = friends_watched_movies(user_data)
+    # get difference between user and friends
+    unique_movies = user_movies - friends_movie_set
+    # store difference in dict
+    diff_dict = [{"title":movie} for movie in unique_movies]
+    return diff_dict
```

Comment:
verrry interesting! now it may not be very efficient since you are rebuilding the entire dictionary again, but I love the idea. If you had done it in a for loop, perhaps the space complexity would be lowered since there would be fewer data structures created.

but I love that you tried a data structure we taught you and made it work! clever!